src/core/legendimageprovider.cpp

```diff
@@ -42,8 +44,23 @@ QPixmap LegendImageProvider::requestPixmap( const QString &id, QSize *size, cons
     for ( int i = 0; i < legendCount; ++i )
     {
       QModelIndex legendIndex = mLayerTreeModel->index( i, 0, layerIndex );
-      if ( idParts.value( 2 ) == mLayerTreeModel->data( legendIndex ) )
-        return mLayerTreeModel->data( legendIndex, Qt::DecorationRole ).value<QPixmap>();
+      if ( idParts.value( 2 ) == mLayerTreeModel->data( legendIndex ).toString() )
+      {
+
+        QPixmap pixmap = mLayerTreeModel->data( legendIndex, Qt::DecorationRole ).value<QPixmap>();
+        if ( pixmap.isNull() )
+        {
+         QIcon icon = mLayerTreeModel->data( legendIndex, Qt::DecorationRole ).value<QIcon>();
+         if ( !icon.isNull() )
+           pixmap = icon.pixmap( 24, 24 );
```

Comment:
This is just they size we get them from QGIS, i.e. no possibility to get them with higher DPI?

Reply:
@m-kuhn , I've only ever stumbled on raster legend nodes outputting QIcons so far, and those are plain color fills.

Reply:
No borders and such?

Reply:
Nope.